### PR TITLE
[DNM] [Sprites Needed] Adds Smithing

### PR DESCRIPTION
**Pull request by Vexylius**
## Ports
https://github.com/Skyrat-SS13/Skyrat-tg/pull/11295
https://github.com/Skyrat-SS13/Skyrat-tg/pull/10312 (partially)
and takes some sprites from
https://github.com/fortune13-ss13/thewasteland/pull/133

## About The Pull Request

This PR adds a NEW and EXCITING mechanic of forging your stuff in a forge!
In order to become a space blacksmith, you will need some tools:
* Forging Billows, crafted out of 3 wood and 2 leather with a help of a hatchet(from autolathe for example),
* Forging Tongs, crafted out of 3 ~~iron~~ metal rods with a help of a welder and wirecutters,
* Forging Hammer, crafted out of 3 ~~iron~~ metal rods with a help of a welder.

...and a proper workstation:
* Anvil - out of 10 sheets of ~~iron~~ metal (sheetcrafting)
* Forge - out of another 10 sheets or ~~iron~~ metal (sheetcrafting)
* Water Basin - out of 5 wooden planks (sheetcrafting)
* Forging Workbench - out of 5 wooden planks (sheetcrafting)

Selected sprites:
image

You start by adding some wood into your forge and heating it up with billows.
Then you take an iron rod in your tongs and insert it into the hot forge. In there you select an item that you want to craft. Once the heated, raw product pops out of the forge, drop your rods stack (tongs take the entire stack of rods, drop it using tongs in hand/press Z while holding 'em) and move that item with your tongs to the anvil. Hit it several times, cool it down in the basin when it's ready. Then, move it to the crafting bench, add whatever's left to finish it and slam the workbench with a hammer one last time. Voilà, you are a blacksmith now.

Also a protip: you can make a crank cell by slapping a coil on any cell. It discharges on its own, but you can use it in hand to charge it by cranking it.

Tongs, forging hammer, water basin, crank cell sprites are made by me :)
Also made the sounds of forge failure by mixing some exiting game files, hope you enjoy it.
## Why It's Good For The Game
This PR adds a fun new mechanic of crafting low-tech weaponry that fits the server.

## Changelog
:cl:
add: Added Smithing
/:cl:

### CODE REVIEW

**Comment by triplezeta** [comment]
sprites :(

**Comment by Vexylius** [comment]
it's a draft, gonna change the sprites later or something.
If anyone wants to help with those sprites my DMs are open

File: code/modules/forging/forging_items.dm [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	name = "primitive forging tongs"
+	work_time = 3 SECONDS
+
+/obj/item/forging/tongs/attack_self(mob/user, modifiers)
+	. = ..()
+	var/obj/search_obj = locate(/obj) in contents
+	if(search_obj)
+		search_obj.forceMove(get_turf(src))
+		icon_state = "tong_empty"
+		return
```
Comment (Fikou) [inline code/modules/forging/forging_items.dm:19]: i'd rather we handle this as a variable than use locate 20 different times.

File: code/modules/forging/forged_weapons.dm [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	name = "reagent staff"
+	desc = "A staff that can be imbued with a reagent. It has a very soft swing."
+	force = 0
+	icon_state = "staff"
+	item_state = "staff"
+	throwforce = 0
+	slot_flags = ITEM_SLOT_BACK
+	w_class = WEIGHT_CLASS_NORMAL
+	resistance_flags = FIRE_PROOF
+/obj/item/forging/reagent_weapon/staff/attack(mob/living/M, mob/living/user, params)
```
Comment (Fikou) [inline code/modules/forging/forged_weapons.dm]: newline here

File: code/game/objects/items/stacks/rods.dm [hunk 3]
```diff
@@ -71,7 +71,14 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
...
 				user.put_in_hands(new_item)
-
+	if(istype(W, /obj/item/forging/tongs))
+		var/obj/searchObj = locate(/obj) in W.contents
+		if(searchObj)
+			to_chat(user, "<span class='warning'>The tongs are already holding something!</span>")
+			return
+		forceMove(W)
+		W.icon_state = "tong_full"
+		return
```
Comment (Fikou) [inline code/game/objects/items/stacks/rods.dm:81]: this should be handled on the tongs' afterattack

File: code/modules/forging/anvil.dm [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	icon_state = "anvil_empty"
+
+	anchored = TRUE
+	density = TRUE
+
+/obj/structure/anvil/Initialize()
+	. = ..()
+
+/obj/structure/anvil/attackby(obj/item/I, mob/living/user, params)
+	var/obj/item/forging/incomplete/search_incomplete_src = locate(/obj/item/forging/incomplete) in contents
```
Comment (Fikou) [inline code/modules/forging/anvil.dm:13]: again, variable, dont look through contents

File: code/modules/forging/anvil.dm [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	name = "anvil"
+	desc = "An object with the intent to hammer metal against. One of the most important parts for forging an item."
+	icon = 'icons/obj/forge_structures.dmi'
+	icon_state = "anvil_empty"
+
+	anchored = TRUE
+	density = TRUE
+
+/obj/structure/anvil/Initialize()
+	. = ..()
```
Comment (Fikou) [inline code/modules/forging/anvil.dm]: dont copy parents proc without changing anything

File: code/modules/forging/anvil.dm [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	if(istype(I, /obj/item/forging/tongs))
+		var/obj/item/forging/forge_item = I
+		if(forge_item.in_use)
+			to_chat(user, "<span class='warning'>You cannot do multiple things at the same time!</span>")
+			return
+		var/obj/item/forging/incomplete/search_incomplete_item = locate(/obj/item/forging/incomplete) in I.contents
+		if(search_incomplete_src && !search_incomplete_item)
+			search_incomplete_src.forceMove(I)
+			icon_state = "anvil_empty"
+			I.icon_state = "tong_full"
```
Comment (Fikou) [inline code/modules/forging/anvil.dm:49]: this should be handled by update_icon_state on both objects

File: code/modules/forging/anvil.dm [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+/obj/structure/anvil/Initialize()
+	. = ..()
+
+/obj/structure/anvil/attackby(obj/item/I, mob/living/user, params)
+	var/obj/item/forging/incomplete/search_incomplete_src = locate(/obj/item/forging/incomplete) in contents
+	if(istype(I, /obj/item/forging/hammer) && search_incomplete_src)
+		playsound(src, 'sound/misc/forge.ogg', 50, TRUE)
+		if(COOLDOWN_FINISHED(search_incomplete_src, heating_remainder))
+			to_chat(user, "<span class='warning'>You mess up, the metal was too cool!</span>")
+			search_incomplete_src.times_hit -= 3
```
Comment (Fikou) [inline code/modules/forging/anvil.dm]: use defines

File: code/modules/forging/anvil.dm [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+			return TRUE
+		if(COOLDOWN_FINISHED(search_incomplete_src, striking_cooldown))
+			var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * search_incomplete_src.average_wait
+			COOLDOWN_START(search_incomplete_src, striking_cooldown, skill_modifier)
+			search_incomplete_src.times_hit++
+			to_chat(user, "<span class='notice'>good hit!</span>") //ReplaceWithBalloonAlertLater
```
Comment (Fikou) [inline code/modules/forging/anvil.dm]: should be capitalized if its not a balloon alert

File: code/modules/forging/forged_weapons.dm [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+	if(istype(attacking_item, /obj/item/forging/hammer))
+		var/obj/item/forging/hammer/attacking_hammer = attacking_item
+		var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * attacking_hammer.work_time
+		while(obj_integrity < max_integrity)
+			if(!do_after(user, skill_modifier, src))
+				return
+			var/fixing_amount = min(max_integrity - obj_integrity, 5)
+			obj_integrity += fixing_amount
+			user.mind.adjust_experience(/datum/skill/smithing, 5) //useful heating means you get some experience
+			to_chat(user, "<span class='notice'>partially repaired!</span>") //ReplaceWithBalloonAlertLater
```
Comment (Fikou) [inline code/modules/forging/forged_weapons.dm]: same here

File: code/modules/forging/forging_items.dm [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+	icon_state = "shovelhead"
+	spawning_item = /obj/item/shovel/reagent_weapon
+
+/obj/item/forging/coil
+	name = "coil"
+	desc = "A simple coil, comprised of coiled iron rods."
+	icon_state = "coil"
+
+/obj/item/stock_parts/cell/attackby(obj/item/attacking_item, mob/user, params)
+	if(istype(attacking_item, /obj/item/forging/coil))
```
Comment (Fikou) [inline code/modules/forging/forging_items.dm]: use early returns here

File: code/modules/forging/forged_weapons.dm [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	new /obj/item/forging/complete/plate(get_turf(src))
+
+/obj/item/shield/riot/buckler/reagent_weapon/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>Using a hammer on [src] will repair its damage!</span>"
+
+/obj/item/shield/riot/buckler/reagent_weapon/attackby(obj/item/attacking_item, mob/user, params)
+	if(obj_integrity >= max_integrity)
+		return ..()
+	if(istype(attacking_item, /obj/item/forging/hammer))
```
Comment (Fikou) [inline code/modules/forging/forged_weapons.dm]: early return here

File: code/modules/forging/forge.dm [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+	check_in_use() //plenty of weird bugs, this should hopefully fix the in_use bugs
+
+/obj/structure/forge/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	var/user_smithing_skill = user.mind.get_skill_level(/datum/skill/smithing)
+	var/previous_level = forge_level
+	if(user_smithing_skill <= SKILL_LEVEL_NOVICE)
+		to_chat(user, "<span class='notice'>[src] requires you to be more experienced!</span>")
+		return
+	if(user_smithing_skill >= SKILL_LEVEL_APPRENTICE)
```
Comment (Fikou) [inline code/modules/forging/forge.dm]: what are these even. why does touching the forge fully upgrade its stuff without you having to do anything

File: code/modules/forging/forge.dm [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+			return
+		in_use = TRUE
+		if(!forge_fuel_strong && !forge_fuel_weak) //if there isnt any fuel, no billow use
+			fail_message(user, "You cannot use [forge_item] without some sort of fuel in [src]!")
+			return
+		if(forge_temperature >= MAX_FORGE_TEMP) //we don't want the "temp" to overflow or something somehow
+			fail_message(user, "You can't heat [src] to be any hotter!")
+			return
+		to_chat(user, "<span class='warning'>You start to pump [forge_item] into [src]...</span>")
+		while(forge_temperature < 91)
```
Comment (Fikou) [inline code/modules/forging/forge.dm]: these should be defines, not magic numbers

**Review by Fikou — CHANGES_REQUESTED** [review]
first look, i really am not a fan of reagent weapons/armor but ill let someone else decide on that

File: code/modules/forging/forged_weapons.dm (hunk 9 above)
Comment (Fikou) [reply code/modules/forging/forged_weapons.dm]: and throughout the rest

**Comment by Fikou** [comment]
Also please credit the original creators as I think this is a port?

File: code/modules/forging/forged_wearables.dm [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	desc = "A tiny ring, sized to wrap around a finger. It is possible to imbue this ring."
+	icon_state = "ringsilver"
+	item_state = "sring"
+	mob_overlay_icon = "sring"
+
+/obj/item/clothing/gloves/ring/reagent_clothing/Initialize()
+	. = ..()
+	AddComponent(/datum/component/reagent_clothing, ITEM_SLOT_GLOVES)
+
+/obj/item/clothing/neck/kink_collar/reagent_clothing
```
Comment (triplezeta) [inline code/modules/forging/forged_wearables.dm]: 🤨

**Comment by Apogee-dev** [comment]
ok i strongly suggest looking at vgstation's blacksmithing-related sprites to replace some of these codersprites. there's an assortment of blacksmithing-related equipment as well as a number of smithed weapons you could steal sprites for and while they stick out pretty bad on vg their style would mesh pretty well with our other sprites over here

**Comment by ZephyrTFA** [comment]
It's not stealing if you ask and/or credit them correctly

**Comment by Latentish** [comment]
This would actually be a pretty cool thing for the icecropolis I think.

**Comment by Latentish** [comment]
> This PR has been inactive for long enough to be automatically marked as stale. This means it is at risk of being auto closed in ~ 14 days, please address any outstanding review items and ensure your PR is finished, if these are all true and you are auto-staled anyway, you need to actively ask maintainers if your PR will be merged. Once you have done any of the previous actions then you should request a maintainer remove the stale label on your PR, to reset the stale timer. If you feel no maintainer will respond in that time, you may wish to close this PR youself, while you seek maintainer comment, as you will then be able to reopen the PR yourself

-NOOOOOOOOO

**Comment by Vexylius** [comment]
I honestly forgot about that, damn

**Pull request closed without merging**